src/adapter/src/coord/command_handler.rs

```diff
@@ -302,9 +324,21 @@ impl Coordinator {
                 self.begin_session_for_statement_logging(&conn);
                 self.active_conns.insert(conn_id.clone(), conn);
 
-                // Note: Do NOT await the notify here, we pass this back to whatever requested the
-                // startup to prevent blocking the Coordinator on a builtin table update.
-                let notify = self.builtin_table_update().defer(vec![update]);
+                let notify = if self.controller.read_only() {
+                    self.buffered_builtin_table_updates
+                        .as_mut()
+                        .expect("in read-only mode")
+                        .push(update);
+
+                    futures::future::ready(()).boxed()
+                } else {
+                    // Note: Do NOT await the notify here, we pass this back to
+                    // whatever requested the startup to prevent blocking the
+                    // Coordinator on a builtin table update.
+                    let notify = self.builtin_table_update().defer(vec![update]);
+
+                    notify
+                };
```

Comment:
Yes, and yes!

Nikhil has this longer-term idea that the new read-only env _does_ create new builtin collections, in which it writes things down, and so the old env can't see them. But we said we wouldn't do that in the first versions. I'm skeptical we can do it even mid-term, because the cut over seems difficult.